reading_comprehension/qanet_encoder.py

```diff
@@ -0,0 +1,345 @@
+        stochastically dropped according to its layer dropout probability.
+    attention_dropout_prob : ``float``, optional, (default = 0)
+        The dropout probability for the attention distributions in the attention layer.
+    """
+
+    def __init__(self,
+                 input_dim: int,
+                 hidden_dim: int,
+                 attention_projection_dim: int,
+                 feedforward_hidden_dim: int,
+                 num_blocks: int,
+                 num_convs_per_block: int,
+                 conv_kernel_size: int,
+                 num_attention_heads: int,
+                 use_positional_encoding: bool = True,
+                 dropout_prob: float = 0.1,
+                 layer_dropout_undecayed_prob: float = 0.1,
+                 attention_dropout_prob: float = 0) -> None:
+        super(StackedQANetEncoder, self).__init__()
+
+        self._input_projection_layer = None
+
+        if input_dim != hidden_dim:
+            self._input_projection_layer = torch.nn.Linear(input_dim, hidden_dim)
+
+        self._encoder_blocks: List[QANetEncoderBlock] = []
+        for block_idx in range(num_blocks):
+            encoder_block = QANetEncoderBlock(hidden_dim,
+                                              hidden_dim,
+                                              attention_projection_dim,
+                                              feedforward_hidden_dim,
+                                              num_convs_per_block,
+                                              conv_kernel_size,
+                                              num_attention_heads,
+                                              use_positional_encoding,
+                                              dropout_prob,
+                                              layer_dropout_undecayed_prob,
+                                              attention_dropout_prob)
+            self.add_module(f"encoder_block_{block_idx}", encoder_block)
+            self._encoder_blocks.append(encoder_block)
+
+        self._input_dim = input_dim
+        self._output_dim = hidden_dim
+
+    @overrides
+    def get_input_dim(self) -> int:
+        return self._input_dim
+
+    @overrides
+    def get_output_dim(self) -> int:
+        return self._output_dim
+
+    @overrides
+    def is_bidirectional(self) -> bool:
+        return False
+
+    @overrides
+    def forward(self, inputs: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:  # pylint: disable=arguments-differ
+        if self._input_projection_layer is not None:
+            inputs = self.dropout(self._input_projection_layer(inputs))
+
+        output = inputs
+        for encoder_block in self._encoder_blocks:
+            output = encoder_block(output, mask)
+
+        return output
+
+
+@Seq2SeqEncoder.register("memory_efficient_multi_head_self_attention")
+class MemoryEfficientMultiHeadSelfAttention(MultiHeadSelfAttention):
```

Comment:
These modifications should just be submitted upstream, I think.  It's fine to have them in here for now, though, until it gets merged upstream.